Client/Frontend/Browser/Playlist/Cache/PlaylistCacheLoader.swift

```diff
@@ -0,0 +1,780 @@
+                return
+            }
+            
+            // We have to create an AVURLAsset here to determine if the item is playable
+            // because otherwise it will add an invalid item to playlist that can't be played.
+            // IE: WebM videos aren't supported so can't be played.
+            // Therefore we shouldn't prompt the user to add to playlist.
+            if let url = URL(string: item.src), !AVURLAsset(url: url).isPlayable {
+                timeout?.cancel()
+                timeout = nil
+                webLoader?.handler(nil)
+                self.webLoader?.tab.webView?.loadHTMLString("<html><body>PlayList</body></html>", baseURL: nil)
+                return
+            }
+            
+            if !playlistItems.contains(item.src) {
+                playlistItems.insert(item.src)
+                
+                timeout?.cancel()
+                timeout = nil
+                webLoader?.handler(item)
+            }
+            
+            // This line MAY cause problems.. because some websites have a loading delay for the source of the media item
+            // If the second we receive the src, we reload the page by doing the below HTML,
+            // It may not have received all info necessary to play the item such as MetadataInfo
+            // For now it works 100% of the time and it is safe to do it. If we come across such a website, that causes problems,
+            // we'll need to find a different way of forcing the WebView to STOP loading metadata in the background
+            DispatchQueue.main.async {
+                self.webLoader?.tab.webView?.loadHTMLString("<html><body>PlayList</body></html>", baseURL: nil)
+            }
+        }
+    }
+    
+    private struct PageInfo: Codable {
+        let pageLoad: Bool
+        
+        public static func from(message: WKScriptMessage) -> PageInfo? {
+            if !JSONSerialization.isValidJSONObject(message.body) {
+                return nil
+            }
+            
+            do {
+                let data = try JSONSerialization.data(withJSONObject: message.body, options: [.fragmentsAllowed])
+                return try JSONDecoder().decode(PageInfo.self, from: data)
+            } catch {
+                log.error("Error Decoding PageInfo: \(error)")
+            }
+            
+            return nil
+        }
+    }
+}
+
+extension PlaylistWebLoader: WKNavigationDelegate {
+    // Recognize an Apple Maps URL. This will trigger the native app. But only if a search query is present. Otherwise
+    // it could just be a visit to a regular page on maps.apple.com.
+    fileprivate func isAppleMapsURL(_ url: URL) -> Bool {
+        if url.scheme == "http" || url.scheme == "https" {
+            if url.host == "maps.apple.com" && url.query != nil {
+                return true
+            }
+        }
+        return false
+    }
```

Comment:
Why do we have all these duped functions from BVC?? What are the diversion points that make this necessary?

Don't think this is an acceptable solution unless there is a huge diversion between logic. Any changes we make in BVC's implementation will likely be forgotten and not tracked here

Reply:
We need the BVC duplicate code because it needs it for adblocking and other fingerprinting and blocking logic.. but at the same time it does not need the javascript alerts and apple maps, universal links, mail, tel, DownloadHelper, PassKit, etc.. logic.
Without this, ad-blocking doesn’t work and refactoring BVC to not be so dependant is going to take a huge amount of work.

We should refactor both BVC and this at once when playlist is merged.